docs/design/sql/14-stream-to-stream-join.md

```diff
@@ -0,0 +1,232 @@
+| s2_wm(6)                   |  
+| [NULL, s2_event(7)]`[**]`  |  
+| s2_wm(8)                   |  
+
+__Table 4__
+
+#### Processor algorithm description
+
+Consider having two input streams __S1__ and __S2__. Let's define the schema for __S1__ and __S2__ as
+
+```sql
+CREATE MAPPING S1 (id BIGINT, payload BIGINT, s1_time TIMESTAMP) TYPE Stream
+CREATE MAPPING S2 (id BIGINT, payload VARCHAR, s2_time TIMESTAMP) TYPE Stream
+```
+
+Consider the following query:
+
+```sql
+> SELECT * FROM IMPOSE_ORDER(TABLE(S1), DESCRIPTOR(s1_time), INTERVAL '1' SECOND) AS s1
+  INNER JOIN 
+  SELECT * FROM IMPOSE_ORDER(TABLE(S2), DESCRIPTOR(s2_time), INTERVAL '1' SECOND) AS s2
+  ON s2.s2_time BETWEEN s1.s1_time AND s1.s1_time + INTERVAL '2' SECOND 
+```
+
+1. Perform query analysis, detect timestamp column from both input stream schemas
+2. Prepare two buffers : B0 to store input events from ordinal 0 (stream S1) and B1 to store input events from ordinal
+   1 (stream S2).
+3. Receive event E from the ordinal.
+    1. If received event is watermark:
+        1. If watermark received from ordinal 0 clean all expired events in B0.
+        2. Else, clean all expired events in B1.
+        3. Emit watermark event to the outbox.
+    2. Else:
+        1. Extract timestamp from event E.
+        2. If extract attempt was failed - throw an exception.
+        3. Store E to the buffer B0, if received from ordinal 0, store to B1 otherwise.
+        4. For each event in 'parallel' buffer
+            1. Test the JOIN condition with received timestamps.
+            2. Perform JOIN operation for each event in 'parallel' buffer, perform
+                1. If test was not successful, retry step 3.
+            3. If the join type is `OUTER JOIN`, we should fill empty side (no input events received) with NULL.
+            4. Emit joined event.
+
+#### Watermarks
+
+JOIN processor would consume stream events from two input streams and join them with regard to JOIN condition. Result
+will not be emitted in case one stream is producing events while the other produces no events for a period of time. We
+have limited amount of memory to store input events. In that case, we require watermarks to used to drop late items and
+to drop overdue items. For that, Hazelcast SQL engine supports `IMPOSE_ORDER` function to add watermarks to stream.
+Non-watermarked streams are not allowed to be joined.
+
+Separate JOIN processor should also emit watermarks. For that, we should extend `Watermark` class in such manner:
+
+```java
+
+@SuppressWarnings("ClassCanBeRecord")
+public final class Watermark implements BroadcastItem {
+    private final int key;        // <-- new field
+    private final long timestamp;
+    // ...
+}
+```
+
+`[*]` __Note__: each watermark event should stay as a separate event in joined stream.
+
+_Q: Should we support only timestamps as JOIN condition?_
+**Sasha's proposition: Yes. Non-timestamp JOIN condition is exceptionally non-trivial to support.**
```

Comment:
Does this mean that the condition must contain a condition on timestamp **only** or that it must contain a condition on timestamp and **may contain** a condition on other fields?

Reply:
My proposition that it **must have only** timestamp condition. Will clarify it in doc.

Reply:
Would be good to update the example above to make this clear.

But, doesn't that reduce the usability of the feature? I suppose you can add a `WHERE` clause to filter out items with e.g. same `order_id`, but you can do that only on smaller data sets.

Reply:
The two biggest problems with a `JOIN` condition without timestamps are how to figure out how long we should store events for a JOIN, and how to extract information about the event store time from such condition.

It doesn't mean that you can't apply `WHERE` clause - it's a bit different from `JOIN` condition.

I personally would like to see at least limited usability of the feature than its absence :)

Reply:
I understand that you need the condition on timestamp, but can you in the JOIN clause specify also *additional* condition, e.g. on `order_id` field? 

If not, then you are effectively doing a cartesian product on the records matching the window, right? Or maybe I am missing something. Maybe you should provide a full examples how the results will look like for all the join types and when the records are present/missing in the specified time window.

Side note: if you make the time window in the condition large enough you effectively remove the time condition, you just need to be able to keep all the records, which won't be practical in most real world scenarios.

Reply:
> but can you in the JOIN clause specify also additional condition, e.g. on order_id field?
Good catch, btw. It may eliminate `only` in **must have only** `JOIN` condition requirement. But still, time boundness  must be.